sw/device/lib/dif/dif_kmac_unittest.cc

```diff
@@ -86,5 +90,81 @@ TEST(FunctionNameTest, BadArg) {
   EXPECT_EQ(dif_kmac_function_name_init("", -1, &fn), kDifBadArg);
 }
 
-}  // namespace
+using mock_mmio::MmioTest;
+using mock_mmio::MockDevice;
+using testing::Test;
+
+// Base class for the rest fixtures in this file.
+class KmacTest : public testing::Test, public mock_mmio::MmioTest {
+ public:
+  dif_kmac_t kmac_;
+  dif_kmac_operation_state_t op_state_;
+  static constexpr std::array<uint8_t, 17> kMsg_ = {
+      0xa7, 0x48, 0x47, 0x93, 0x0a, 0x03, 0xab, 0xee, 0xa4,
+      0x73, 0xe1, 0xf3, 0xdc, 0x30, 0xb8, 0x88, 0x15};
+
+  KmacTest() { EXPECT_EQ(dif_kmac_init(dev().region(), &kmac_), kDifOk); }
+
+  /**
+   * Set mmio write expectation for 8 bits data size.
+   *
+   * @param message Buffer with the data.
+   * @param size Len of the buffer.
+   */
+  void setExpectedMessageByte(const uint8_t *message, const size_t size) {
+    for (int i = 0; i < size; ++i) {
+      EXPECT_WRITE8(KMAC_MSG_FIFO_REG_OFFSET, message[i]);
+    }
+  }
+
+  /**
+   * Set mmio write expectation for 32 bits data size considering an alignment
+   * of 32 bits.
+   *
+   * @param message Buffer with the data.
+   * @param size Len of the buffer.
+   */
+  void setExpectedMessageInt32(const uint8_t *message, const size_t size) {
+    // Check if the buffer is unaligned.
+    size_t remaining = size;
+    size_t unalignment = ((uintptr_t)message) % sizeof(uint32_t);
+    if (unalignment) {
+      // write unaligned data in bytes.
+      unalignment = sizeof(uint32_t) - unalignment;
+      setExpectedMessageByte(message, unalignment);
+      message += unalignment;
+      remaining -= unalignment;
+    }
+
+    // Write aligned part of the buffer.
+    while (remaining >= sizeof(uint32_t)) {
+      uint32_t word = 0;
+      memcpy(&word, message, sizeof(uint32_t));
+      EXPECT_WRITE32(KMAC_MSG_FIFO_REG_OFFSET, word);
+      remaining -= sizeof(uint32_t);
+      message += sizeof(uint32_t);
+    }
+    // Check if there still unaligned data in the buffer tail.
+    if (remaining) {
+      // write unaligned data in bytes.
+      setExpectedMessageByte(message, remaining);
+    }
+  }
+};
+constexpr std::array<uint8_t, 17> KmacTest::kMsg_;
```

Comment:
I don't really need this, as you already define it in the class `public` section.

Reply:
The reazon is that, to have a `constexpr` attributte in a class, it sould also be static (image 1), and every `static` attribute must me define and declared (image 2).

image 1
![image](https://user-images.githubusercontent.com/92785679/142882897-564376bc-a4ca-4e95-939f-dbac9fd911c2.png)

Image 2
![image](https://user-images.githubusercontent.com/92785679/142883045-0cbb12c4-ef30-4fcd-814e-04116b001c01.png)

I personally would use `const` rather than `constexpr` in this case, but I following Michaels adivese. How would you do?

Reply:
I think you probably can use it inside your test function as `KmacTest::kMsg_` directly. I think that is probably what I would have went for, what do you think @mundaym ? Anyway, I think this is very minor.

Reply:
Yeah, as @silvestrs says its pretty minor. Personally I would have made it a static class member since it is a compile time constant.

Reply:
Oh wait, that is what you are doing. Yeah, I think this is fine.